opensquirrel/utils/matrix_expander.py

```diff
@@ -0,0 +1,185 @@
+import math
+
+import numpy as np
+
+from opensquirrel.common import Can1
+from opensquirrel.gates import MultiQubitMatrixSemantic, Semantic, SingleQubitAxisAngleSemantic
+
+
+def extract_bits(x: int, bit_indices: [int]) -> int:
+    """
+    Extract the bits at given indices, placing the bits in order from least to most significant.
+    Equivalent to pext instruction.
+
+    Args:
+        x: a bitstring
```

Comment:
> Also, is integer the 'right' type for a bitstring?

Good remark
I've looked into alternatives. Ideally i would be able to access bit #i with [i], but I also need to be able to increment (which is something typically from integers).
Because each bitstring comes from a unitary matrix's row or column index...
I could use an array of bool (probably super inefficient) but it doesn't let me increment it
I could use a Python bytes object, but I think the [] operator accesses a whole byte, not just a bit

Reply:
Wouldn't it be better to translate an input bit string to a convenient representation in the method, perform the operation, and then return a bit string again. I understand that this is likely a helper function (i.e. only used internally and not exposed to the user), but it is confusing to consider the integer 10 to stand for 2, i.e. '0b10'.

Also, in stead of `x`, it is more informative to call the parameter `bitstring`.

Reply:
I think that it is actually relatively common in the literature to identify a quantum ket with its integer representation in base 10.
I mean by that, instead of writing |001010>, people write |10>
and instead of |1111> people write |15>
It is also a very natural and efficient representation for a classical computer.

In this code I am really dealing with _integers_: I get an integer ket and I extract the integer index of the column that is applied in the "small matrix". It turns out that this "extraction" involves bitwise manipulations.

I think your comment might have to do with bitwise manipulations being sometimes tricky to understand. I get that.

I can use strings to write bitstrings, e.g. in Python turn an integer `i` into `bin(i)[2:]`
that is a very inefficient use of storage and doesn't handle, for instance, bit masking and shifting and ... Plus it needs to be converted back and forth to an integer anyway, since I query the corresponding column of the numpy matrix. A nightmare in terms of performance.

An alternative is to create a wrapper class that is a view on the bits of an integer. Honestly I'm not sure it's worth it.

Reply:
I wrote those aux functions as "generic" bit manipulations

I'm wondering if, to make it easier to understand, I can give them names like
`get_small_matrix_column` (for extract_bits)
and `get_large_matrix_row` (for clean + deposit)

what do you think?

Reply:
> I wrote those aux functions as "generic" bit manipulations
> 
> I'm wondering if, to make it easier to understand, I can give them names like `get_small_matrix_column` (for extract_bits) and `get_large_matrix_row` (for clean + deposit)
> 
> what do you think?

Yes that would be good, and you can add to the description that they are basically similar to an extract (PEXT)/deposit (PDEP) operation, respectively.